AsyncDisplayKit/ASNetworkImageNode.mm

```diff
@@ -350,41 +348,56 @@ - (CGSize)calculateSizeThatFits:(CGSize)constrainedSize
 {
   ASDN::MutexLocker l(__instanceLock__);
-  
-  BOOL shouldRenderProgressImages = _shouldRenderProgressImages;
-  ASInterfaceState interfaceState = self.interfaceState;
+  // Getting a result back for a different download identifier, download must not have been successfully canceled
+  if (ASObjectIsEqual(_downloadIdentifier, downloadIdentifier) == NO && downloadIdentifier != nil) {
+    return;
+  }
+  self.image = progressImage;
+  dispatch_async(dispatch_get_main_queue(), ^{
+    // See comment in -displayDidFinish for why this must be dispatched to main
+    self.currentImageQuality = progress;
+  });
+}
 
-  if (!_downloaderFlags.downloaderImplementsSetProgress || _downloadIdentifier == nil) {
+#pragma mark - Private methods, call with lock held
+
+- (void)_updateProgressImageBlockOnDownloaderIfNeeded
+{
+  // If the downloader doesn't do progress, we are done.
+  if (_downloaderFlags.downloaderImplementsSetProgress == NO) {
     return;
   }
 
-  ASImageDownloaderProgressImage progress = nil;
-  if (shouldRenderProgressImages && ASInterfaceStateIncludesVisible(interfaceState)) {
-    __weak __typeof__(self) weakSelf = self;
-    progress = ^(UIImage * _Nonnull progressImage, CGFloat progress, id _Nullable downloadIdentifier) {
-      __typeof__(self) strongSelf = weakSelf;
-      if (strongSelf == nil) {
-        return;
-      }
+  // Read state.
+  BOOL shouldRender = _shouldRenderProgressImages && ASInterfaceStateIncludesVisible(_interfaceState);
+  id oldDownloadIDForProgressBlock = _downloadIdentifierForProgressBlock;
+  id newDownloadIDForProgressBlock = shouldRender ? _downloadIdentifier : nil;
 
-      ASDN::MutexLocker l(strongSelf->__instanceLock__);
-      //Getting a result back for a different download identifier, download must not have been successfully canceled
-      if (ASObjectIsEqual(strongSelf->_downloadIdentifier, downloadIdentifier) == NO && downloadIdentifier != nil) {
-        return;
-      }
-      strongSelf.image = progressImage;
-      dispatch_async(dispatch_get_main_queue(), ^{
-        // See comment in -displayDidFinish for why this must be dispatched to main
-        strongSelf.currentImageQuality = progress;
-      });
-    };
+  // If we're already bound to the correct download, we're done.
+  if (ASObjectIsEqual(oldDownloadIDForProgressBlock, newDownloadIDForProgressBlock)) {
+    return;
+  }
+
+  // Unbind from the previous download.
+  if (oldDownloadIDForProgressBlock != nil) {
+    [_downloader setProgressImageBlock:nil callbackQueue:dispatch_get_main_queue() withDownloadIdentifier:oldDownloadIDForProgressBlock];
   }
-  [_downloader setProgressImageBlock:progress callbackQueue:dispatch_get_main_queue() withDownloadIdentifier:_downloadIdentifier];
+
+  // Bind to the current download.
+  if (newDownloadIDForProgressBlock != nil) {
+    __weak __typeof(self) weakSelf = self;
```

Comment:
Yes but `instancetype` can't be used in this context. It can only be used as the return type for an instance method.